### PR TITLE
Send notifications on simple payment action

### LINKED ISSUE #3010: Notifications: Send notification on simple payment made with permissions
### Story

As a *user*, I want to *receive a notification when a simple payment is made*, so that *I am kept informed*

### Description

When a simple payment is made using permissions decision method, a notification should be created in Magicbell. The recipients of this notification should be everybody who is in the colony who has notifications turned on, and does not have this colony muted, as well as a separate notification for the recipient of the funds.

### This issue is complete when...

- [ ] Notifications are created when a simple payment is made with permissions

### Suggestions (optional)

* In my opinion, this notification should be sent from the block ingestor.

### PR DESCRIPTION
**Pull request by davecreaser**
## Description

Send notifications when a simple payment action is made via permissions.

Most of the work is done in this block ingestor PR: https://github.com/JoinColony/block-ingestor/pull/268, so please also review that.

Notifications should be sent to everybody in the colony who has notifications turned on and has not muted that colony. Another notification should be sent to the recipient of the funds.

## Testing

We need to check that the notifications are sent to the right people. Make sure that you start this testing with a clean environment.

**IMPORTANT:** You need the magicbell api key and secret to be in your .env file.

* Log into the app as Leela, and then run the following query in `localhost:20002` to check that notifications data is made for this user.

```graphql
query MyQuery {
  getUser(id: "0xb77D57F4959eAfA0339424b83FcFaf9c15407461") {
    id
    profile {
      displayName
    }
    notificationsData {
      magicbellUserId
      userAddress
      mutedColonyAddresses
      notificationsDisabled
    }
  }
}
```

Screenshot 2024-09-03 at 17 15 32

* Create a simple payment action with Leela as the recipient.

Screenshot 2024-09-03 at 17 16 49

* Go to the Magicbell dashboard and check that two notifications were sent, each with Leela as the only recipient:

Screenshot 2024-09-03 at 18 25 33
Screenshot 2024-09-03 at 18 25 40

* In an incognito window, log in as Amy (this will make a notifications user for her).

* Now make another simple payment action with Leela as the recipient, and again check the Magicbell dashboard. This time two Notifications should have been made, but one of them should have two recipients (Leela and Amy) while the `mention` notification should still only be going to Leela.

Screenshot 2024-09-03 at 18 29 13
Screenshot 2024-09-03 at 18 29 15

* Run the following mutation to disable notifications for Leela:

```graphql
mutation MyMutation {
  updateNotificationsData(
    input: {userAddress: "0xb77D57F4959eAfA0339424b83FcFaf9c15407461", notificationsDisabled: true}
  ) {
    userAddress
    notificationsDisabled
  }
}
```

* Make another simple payment with Leela as the recipient, and again check the Magicbell dashboard. This time only one notification should have been sent - the colony wide one - with Amy as the only recipient.

Screenshot 2024-09-03 at 18 32 58

* Run the following mutation to mute the planex colony for Amy:

```graphql
mutation MyMutation {
  updateNotificationsData(
    input: {userAddress: "0x9dF24e73f40b2a911Eb254A8825103723E13209C", mutedColonyAddresses: "0x59E0fe5e32A300Ebcb14787c4C198087559C36Ec"}
  ) {
    userAddress
    mutedColonyAddresses
  }
}
```

* Make a final simple payment with Leela as the recipient. This time NO notifications at all should be made.

## Diffs

**New stuff** ✨

* New utils for handling notifications in the block ingestor

**Changes** 🏗

* Changed the model for NotificationsData in amplify

Resolves #3010

### CODE REVIEW

**Comment by davecreaser** [comment]
This is ready for review, please pay special attention to the important note before testing! (And if anyone has any ideas on how best to tackle this please let me know!).

Also please do review the block ingestor PR with this as that is where most of the changes are. Thanks!

**Review by iamsamgibbs — APPROVED** [review]
Great work on this! Followed the testing steps without issue. With just Leela:

Screenshot 2024-09-04 at 08 44 54
Screenshot 2024-09-04 at 08 46 59
Screenshot 2024-09-04 at 08 47 04

With Amy + Leela:

Screenshot 2024-09-04 at 08 48 36
Screenshot 2024-09-04 at 08 48 40

With just Amy:

Screenshot 2024-09-04 at 08 49 18

And got no notifications when both have muted.

I wasn't able to get block-ingestor running via docker even with the env variables added (the issue seems unrelated though)

Screenshot 2024-09-04 at 08 31 50

And I also experienced this issue with the block-ingestor:

Screenshot 2024-09-04 at 08 50 17

This occurred after trying to create a simple payment after both users had notifications muted, however I wasn't able to recreate this and I don't think it's related to any of the changes in this PR.

**Review by Nortsova — CHANGES_REQUESTED** [review]
Hey @davecreaser ! Thanks for this nice PR 👏

Leela has notifications for payment: 
image

After Amy sent a simple payment to Leela, two recipients appeared in the new notification:
image

And the mention notification only went to Leela:
image

After disabling notifications for Leela, it sent a notification only to Amy:
image

Only after muting the colony did I encounter an issue:
image

And the notification was still sent: 
image

So, I think I'll recheck the last step later, as I'm not entirely sure if it's an issue with the current branch or an issue on master.

**Review by mmioana — CHANGES_REQUESTED** [review]
Awesome work on this one @davecreaser 💯 

After running the [block-ingestor](https://github.com/JoinColony/block-ingestor/pull/268) branch locally, can confirm your testing steps work as expected

Before magic bell user data was created for `leela`
Screenshot 2024-09-13 at 10 42 17
After 
Screenshot 2024-09-13 at 10 57 52

Creating a payment for `leela`
Screenshot 2024-09-13 at 10 50 09

Notifications show up in the magic bell dashboard
Screenshot 2024-09-13 at 10 47 47
Screenshot 2024-09-13 at 10 48 22
Screenshot 2024-09-13 at 10 48 29

Opening the app with `amy` and running the query, can confirm the magic bell user is created
Screenshot 2024-09-13 at 10 49 16

Creating another simple payment for `leela`
Screenshot 2024-09-13 at 10 50 09

We get notifications for both `leela` and `amy`, but the mention notification remains only for `leela`
Screenshot 2024-09-13 at 10 50 27
Screenshot 2024-09-13 at 10 50 33

Disable the notifications for `leela`
Screenshot 2024-09-13 at 10 50 53

Create another simple payment
Screenshot 2024-09-13 at 10 51 44

Only one notification for `amy` gets created
Screenshot 2024-09-13 at 10 51 58

Disabling notifications also for `amy` running the query you provided
Screenshot 2024-09-13 at 10 52 28
However realised, the colony address should have been updated, so here is the query run again - @iamsamgibbs maybe this is why you still got notifications for `amy`
Screenshot 2024-09-13 at 10 52 47

Created another simple payment
Screenshot 2024-09-13 at 10 55 07

And no more notifications are send out
Screenshot 2024-09-13 at 10 55 20


@davecreaser the only reason I'm rejecting this is due to the fact that by running the `dev` environment for the colonyCDapp, the block ingestor wouldn't start no matter what, so I had to run the `dev` environment for the block-ingestor repository to be able to test this 
Here is the error 🔴 
Screenshot 2024-09-13 at 10 27 43

Also noticed after disabling the notifications for `leela` and created a simple payment, I got this error and had to restart the block-ingestor
Screenshot 2024-09-13 at 10 54 12

So, if this errors appeared only due to my local environment, I'm happy to approve your PR 👍 
Again super nice work 🥇

File: docker/colony-cdapp-dev-env-block-ingestor [hunk 1]
```diff
@@ -28,6 +28,10 @@ RUN git checkout $BLOCK_INGESTOR_HASH
 # need to add it after the repo has been cloned, otherwise cloning will fail
 ADD docker/files/block-ingestor/env.base /colonyCDappBackend/block-ingestor/.env
 
+# Temporarily mount the .env file from the root of this repo to pass the required magicbell env vars
+# to the block ingestor. You must ensure these two env vars exist in your env file for this to work properly.
+RUN --mount=type=bind,source=/.env,target=/tmpMount/.env grep -e MAGICBELL_API_KEY -e MAGICBELL_API_SECRET /tmpMount/.env >> /colonyCDappBackend/block-ingestor/.env
+
```
Comment (iamsamgibbs) [inline docker/colony-cdapp-dev-env-block-ingestor:34]: Haven't review this yet, but this looks very handy.

Comment (davecreaser) [reply docker/colony-cdapp-dev-env-block-ingestor:34]: It did take a while to figure out this docker black magic 😅 Should be useful for passing any env vars needed from the cdapp to the other apps though

**Review by bassgeta — APPROVED** [review]
Notification data is created for leela:
image
THen after making a simple payment:
image
Then amy's data also gets created :heavy_check_mark: 
image
Muted notifications as `leela`:
image
amy:
image
leela:
image

Nice going! I see that the number of unread notifications doesn't automatically update, but we shouldn't be solving that here anyways.

let's get it merged :sunglasses:

**Review by mmioana — APPROVED** [review]
Awesome work @davecreaser you nailed this one 💪 

Before `leela`s login into the colony
Screenshot 2024-09-20 at 09 51 41
After, the notifications data was created
Screenshot 2024-09-20 at 09 51 56
Creating a simple payment
Screenshot 2024-09-20 at 09 53 16
Both the simple payment and mention notifications got created
Screenshot 2024-09-20 at 09 53 52
Screenshot 2024-09-20 at 09 53 59

Now it's time to add `amy` in the loop
Screenshot 2024-09-20 at 09 54 42

Creating another simple payment
Screenshot 2024-09-20 at 09 55 08

Now a mention notification for `leela` and a simple payment notification for both `leela` and `amy` are created
Screenshot 2024-09-20 at 09 55 27
Screenshot 2024-09-20 at 09 55 33
Muting notifications for `leela`
Screenshot 2024-09-20 at 09 55 49

Creating another simple payment
Screenshot 2024-09-20 at 09 56 10

Only a simple payment notification for `amy` get created
Screenshot 2024-09-20 at 09 56 28

Muting the notifications for `amy` within `planex`
Screenshot 2024-09-20 at 09 57 17
Screenshot 2024-09-20 at 09 57 42

Creating another simple payment
Screenshot 2024-09-20 at 09 58 23

And there are **no** new notifications 🌟 
Screenshot 2024-09-20 at 09 58 39

Great and clean work 👏

**Pull request merged**